model/postgres/metal/postgres_resource.rb

```diff
@@ -10,5 +10,17 @@ def metal_upgrade_candidate_server
         .select { |server| server.vm.vm_storage_volumes.filter { it.boot }.any? { it.boot_image.version >= UPGRADE_IMAGE_MIN_VERSIONS[target_version] } }
         .max_by(&:created_at)
     end
+
+    # If the server is in leaseweb, we don't have multiple DCs anyway.
+    # The first element is the list of vm_hosts to exclude from the new server,
+    # which means the dc of the existing servers for metal
+    # The second element is the list of availability zones to exclude from the new server,
+    # which is empty for metal.
+    # The third element is the availability zone of the representative server,
+    # which is nil for metal.
+    def metal_new_server_exclusion_filters
+      return [[], [], nil] if Config.allow_unspread_servers || location.provider == HostProvider::LEASEWEB_PROVIDER_NAME
+      [VmHost.where(data_center: servers.map { it.vm.vm_host.data_center }.uniq).map(&:id), [], nil]
```

Comment:
2N+1 query issue. Possible fix:

```suggestion
      hosts_ds = VmHost
        .where(id: Vm.where(id: servers_dataset.select(:vm_id))
      exclude_host_ids = VmHost
        .where(data_center: hosts_ds.select(:data_center)
        .distinct
        .select_map(:id)
      [exclude_host_ids, [], nil]
```

We should probably change the allocator to accept an `exclude_data_center` argument so you could just pass the data centers in use (or a dataset like `hosts_ds.select(:data_center)` as shown above), avoiding the need to retrieve the ids of all hosts to exclude.